ENH Add dtype preservation to `LocalOutlierFactor`

jjerphan opened the pull request:
#### Reference Issues/PRs

Partially addresses https://github.com/scikit-learn/scikit-learn/issues/22881
Precedes #22590

#### What does this implement/fix? Explain your changes.

This makes `LocalOutlierFactor` preserves inputs dtype, in particular `np.float32` dtyped inputs.

This also parametrizes tests from `test_lof.py` to run on np.float32 datasets.

jjerphan committed:
TST Adapt test_lof.py to test implementations on 32bit datasets

jjerphan committed:
Merge branch 'main' into tst/test_lof-32bit

jjerphan committed:
TST Use global_dtype

jeremiedbb commented on sklearn/neighbors/tests/test_lof.py:
```diff
@@ -30,9 +30,12 @@
+def test_lof(global_dtype):
     # Toy sample (the last two samples are outliers):
-    X = [[-2, -1], [-1, -1], [-1, -2], [1, 1], [1, 2], [2, 1], [5, 3], [-4, 2]]
+    X = np.asarray(
+        [[-2, -1], [-1, -1], [-1, -2], [1, 1], [1, 2], [2, 1], [5, 3], [-4, 2]],
+        dtype=global_dtype,
+    )
```
it changes the original test. before it was an array of int. Now it's either float32 or float64

jjerphan replied:
Following up on this: we still want to convert, right?

jeremiedbb replied:
Yes I guess

jeremiedbb commented on sklearn/neighbors/tests/test_lof.py:
```diff
@@ -44,14 +47,15 @@ def test_lof():
     # Generate train/test data
     rng = check_random_state(2)
-    X = 0.3 * rng.randn(120, 2)
+    X = 0.3 * rng.randn(120, 2).astype(global_dtype)
```
```suggestion
    X = 0.3 * rng.randn(120, 2).astype(global_dtype, copy=False)
```
and elsewhere

jeremiedbb commented on sklearn/neighbors/tests/test_lof.py:
```diff
@@ -69,9 +73,9 @@ def test_lof_performance():
+def test_lof_values(global_dtype):
     # toy samples:
-    X_train = [[1, 1], [1, 2], [2, 1]]
+    X_train = np.asarray([[1, 1], [1, 2], [2, 1]], dtype=global_dtype)
```
same issue, originally it was an array of int

jeremiedbb commented on sklearn/neighbors/tests/test_lof.py:
```diff
@@ -89,12 +93,12 @@ def test_lof_values():
 
 
-def test_lof_precomputed(random_state=42):
+def test_lof_precomputed(global_dtype, random_state=42):
```
this test needs to use assert_allclose

jeremiedbb replied:
Currently it's using `assert_array_almost_equal` which should be replaced by `assert_allclose`

jeremiedbb commented on sklearn/neighbors/tests/test_lof.py:
```diff
@@ -127,23 +131,22 @@ def test_n_neighbors_attribute():
 
-def test_score_samples():
-    X_train = [[1, 1], [1, 2], [2, 1]]
+def test_score_samples(global_dtype):
```
This test should use assert_allclose

jeremiedbb replied:
When comparing floats, we should always use assert_allclose

jjerphan committed:
TST Do not copy on when casting

jeremiedbb commented on sklearn/neighbors/tests/test_lof.py:
```diff
@@ -44,14 +47,15 @@ def test_lof():
     clf = neighbors.LocalOutlierFactor(contamination=0.25, n_neighbors=5).fit(X)
-    assert_array_equal(clf._predict(), 6 * [1] + 2 * [-1])
-    assert_array_equal(clf.fit_predict(X), 6 * [1] + 2 * [-1])
+    expected_predictions = np.asarray(6 * [1] + 2 * [-1], dtype=global_dtype)
```
Do you need that ? assert_allclose should do that by itself iirc

jjerphan replied:
This was to remove duplication.

jeremiedbb replied:
What I meant is
```suggestion
    expected_predictions = 6 * [1] + 2 * [-1]
```

jeremiedbb commented on sklearn/neighbors/tests/test_lof.py:
```diff
@@ -44,14 +47,15 @@ def test_lof():
     # Generate train/test data
     rng = check_random_state(2)
-    X = 0.3 * rng.randn(120, 2)
+    X = 0.3 * rng.randn(120, 2).astype(global_dtype, copy=False)
     X_train = X[:100]
 
     # Generate some abnormal novel observations
```
X_outliers should be cast as well otherwise it will always be float64 and thus X_test also

jeremiedbb commented on sklearn/neighbors/tests/test_lof.py:
```diff
@@ -69,9 +73,9 @@ def test_lof_performance():
 
 
-def test_lof_values():
+def test_lof_values(global_dtype):
```
This test should use assert_allclose

jjerphan committed:
Address review comments

Co-authored-by: Jérémie du Boisberranger <jeremiedbb@users.noreply.github.com>

jjerphan committed:
fixup! Address review comments

jeremiedbb reviewed (approved):
LGTM

glemaitre commented on sklearn/neighbors/tests/test_lof.py:
```diff
@@ -44,18 +48,21 @@ def test_lof():
-    assert_array_equal(clf.fit_predict(X), 6 * [1] + 2 * [-1])
+    expected_predictions = 6 * [1] + 2 * [-1]
+    assert_array_equal(clf._predict(), expected_predictions)
+    assert_array_equal(clf.fit_predict(X), expected_predictions)
```
Does `negative_outlier_factor_` been of the same `dtype` that `X`. If this is the case, we should either add a test or add a TODO.

glemaitre commented on sklearn/neighbors/tests/test_lof.py:
```diff
@@ -8,13 +8,14 @@
 from sklearn.metrics import roc_auc_score
 
 from sklearn.utils import check_random_state
 from sklearn.utils._testing import assert_array_almost_equal
```
It remains a single occurence of `assert_array_almost_equal`. We can use `assert_allclose` for it and remove this import.

glemaitre commented on sklearn/neighbors/tests/test_lof.py:
```diff
@@ -8,13 +8,14 @@
 from sklearn.utils import check_random_state
 from sklearn.utils._testing import assert_array_almost_equal
+from sklearn.utils._testing import assert_allclose
+from sklearn.utils._testing import assert_array_equal
```
should we use numpy here?

glemaitre replied:
Up I misread the issue. This is actually the opposite.

glemaitre reviewed (commented):
It could be worth to add the `global_dtype` for the `test_predicted_outlier_number` test, I think.

jjerphan committed:
Merge branch 'main' into tst/test_lof-32bit

jjerphan committed:
Review comments

Co-authored-by: Guillaume Lemaitre <g.lemaitre58@gmail.com>

jjerphan commented on sklearn/neighbors/tests/test_lof.py:
```diff
@@ -187,7 +193,8 @@ def test_novelty_training_scores():
 
-    assert_array_almost_equal(scores_1, scores_2)
+    assert_allclose(scores_1, scores_2)
+    assert scores_1.dtype == scores_2.dtype == global_dtype
```
Negative outlier factors are computed from knn distances which are returned in float64.

Must they be float32?

ogrisel replied:
If this is just a scalar value and it does not impact the dtype of the prediction function then it does not matter much.

So if they are naturally computed as float64 then so be it.

jjerphan committed:
Merge branch 'main' into tst/test_lof-32bit

jjerphan committed:
Preserve input dtype for LOF

Tests currently fails because X can be a list of list of numbers
and thus not have a `dtype` attribute.

See this RFC for discussions:
https://github.com/scikit-learn/scikit-learn/issues/24745

jjerphan committed:
Correctly preserves dtypes and assert this preservation

jjerphan commented on sklearn/neighbors/tests/test_lof.py:
```diff
@@ -127,23 +134,22 @@ def test_n_neighbors_attribute():
+        clf2.score_samples(X_test),
+        clf2.decision_function(X_test) + clf2.offset_,
     )
+    assert_allclose(clf1.score_samples(X_test), clf2.score_samples(X_test))
```
Done as part of eacde86dbc63af1273c1d42e550ba431d599e4fc.

jjerphan commented on sklearn/neighbors/_lof.py:
```diff
@@ -462,7 +468,15 @@ def score_samples(self, X):
         )
-        X_lrd = self._local_reachability_density(distances_X, neighbors_indices_X)
+
+        # To preserves inputs dtype for outputs.
```
```suggestion
        # To preserve inputs dtype for outputs.
```

jjerphan committed:
Fix typo

glemaitre commented on sklearn/neighbors/_lof.py:
```diff
@@ -462,7 +468,15 @@ def score_samples(self, X):
+
+        # To preserve inputs dtype for outputs.
+        # This might downcast distances_X from np.float64 to np.float32.
+        distances_X = distances_X.astype(X.dtype, copy=False)
```
Since `dtype="numeric"` in `check_array`, `X.dtype` will preserve and `distance_X` can be cast to whatever numerical `dtype`.

For `distances_X`, I don't think we want to allow to cast to integral `dtype`.

We probably want to have a safe guard here:

```python
if X.dtype == np.float32:
    distances_X = distances_X.atype(X.dtype, copy=False)
```

glemaitre commented on sklearn/neighbors/_lof.py:
```diff
@@ -291,6 +291,12 @@ def fit(self, X, y=None):
             n_neighbors=self.n_neighbors_
         )
 
+        # This allows preserving input dtype for all attributes.
+        self._distances_fit_X_ = self._distances_fit_X_.astype(
+            self._fit_X.dtype,
+            copy=False,
+        )
```
I assume that we have the same issue here, we don't want to make it lower than 32-bitness precision.

```suggestion
        if self._fit_X.dtype == np.float32:
          self._distances_fit_X_ = self._distances_fit_X_.astype(
              self._fit_X.dtype,
              copy=False,
          )
```

glemaitre commented on sklearn/neighbors/_lof.py:
```diff
@@ -462,7 +468,15 @@ def score_samples(self, X):
         )
-        X_lrd = self._local_reachability_density(distances_X, neighbors_indices_X)
+
+        # To preserve inputs dtype for outputs.
+        # This might downcast distances_X from np.float64 to np.float32.
```
I don't think that we need comments here.

glemaitre commented on sklearn/neighbors/_lof.py:
```diff
@@ -462,7 +468,15 @@ def score_samples(self, X):
+        # This might downcast distances_X from np.float64 to np.float32.
+        distances_X = distances_X.astype(X.dtype, copy=False)
+
+        X_lrd = self._local_reachability_density(
+            distances_X,
+            neighbors_indices_X,
+        )
```
```suggestion
        X_lrd = self._local_reachability_density(
            distances_X, neighbors_indices_X
        )
```

glemaitre reviewed (commented):
I did not look at the test yet but I am a bit confused here. If I understand properly, to be able to use `global_dtype` in the test, we have to make `LocalOutlierFactor` to preserve `dtype` since we make some casting in some parts of the code.

glemaitre commented:
We should be adding the `_more_tags` and define that we preserve `np.float32` and `np.float64` dtype. It will complete some of the test where we are testing the type preservation of some of the attributes.

glemaitre commented on sklearn/neighbors/tests/test_lof.py:
```diff
@@ -71,32 +78,32 @@ def test_lof_performance():
     assert roc_auc_score(y_test, y_pred) > 0.99
```
We should be testing somewhere that `y_pred` is preserving dtype also.

jjerphan replied:
I understand your last message as indicating that we are not needing and assertion for `dtype` preservation because this is done in `test_score_samples`.

glemaitre commented:
Otherwise, I think that the other tests are fine.

jjerphan committed:
Add a guard to only cast np.float32 arrays

Co-authored-by: Guillaume Lemaitre <g.lemaitre58@gmail.com>

jjerphan committed:
Add whats_new entry for 1.2

jjerphan commented:
I've changed the scope of this PR from `TST` to `ENH` as it also makes `LocalOutlierFactor` preserve dtypes.

glemaitre reviewed (commented):
I am thinking that we could isolate the check of the attribute dtype in a separate test. It would be more readable. We could also add a test to check the consistency 32/64 bits.

In some way this is what we try to do in other preservation dtype PRs: https://github.com/scikit-learn/scikit-learn/pull/24714/files#diff-1f2d3a6511fee4ecd34ef77089119895929c26b5f1193493eacdfd4e745cded7R253

jjerphan committed:
TST Add dedicated test for dtype preservation

Co-authored-by: Guillaume Lemaitre <g.lemaitre58@gmail.com>

jjerphan committed:
fixup! TST Add dedicated test for dtype preservation

glemaitre commented:
I would still propose adding a test for the equivalence 32/64 bits regarding the prediction/score functions.

I also find it cleaner to delay the check for the `dtype` in the "preservation" test.

I would propose the following patch:

```diff
diff --git a/sklearn/neighbors/tests/test_lof.py b/sklearn/neighbors/tests/test_lof.py
index 6ef5e145d4..3ccb3bc3ea 100644
--- a/sklearn/neighbors/tests/test_lof.py
+++ b/sklearn/neighbors/tests/test_lof.py
@@ -149,12 +149,6 @@ def test_score_samples(global_dtype):
     clf2_scores = clf2.score_samples(X_test)
     clf2_decisions = clf2.decision_function(X_test)
 
-    assert clf1_scores.dtype == global_dtype
-    assert clf1_decisions.dtype == global_dtype
-
-    assert clf2_scores.dtype == global_dtype
-    assert clf2_decisions.dtype == global_dtype
-
     assert_allclose(
         clf1_scores,
         clf1_decisions + clf1.offset_,
@@ -201,7 +195,6 @@ def test_novelty_training_scores(global_dtype):
     scores_2 = clf_2.negative_outlier_factor_
 
     assert_allclose(scores_1, scores_2)
-    assert scores_1.dtype == scores_2.dtype == global_dtype
 
 
 def test_hasattr_prediction():
@@ -278,3 +271,42 @@ def test_lof_input_dtype_preservation(global_dtype, algorithm, contamination, no
     iso.fit(X)
 
     assert iso.negative_outlier_factor_.dtype == global_dtype
+
+    for method in ("score_samples", "decision_function"):
+        if hasattr(iso, method):
+            y_pred = getattr(iso, method)(X)
+            assert y_pred.dtype == global_dtype
+
+
+@pytest.mark.parametrize("algorithm", ["auto", "ball_tree", "kd_tree", "brute"])
+@pytest.mark.parametrize("novelty", [True, False])
+@pytest.mark.parametrize("contamination", [0.5, "auto"])
+def test_lof_dtype_equivalence(algorithm, novelty, contamination):
+    """Check the equivalence of the results with 32 and 64 bits input."""
+
+    inliers = iris.data[:50]  # setosa iris are really distinct from others
+    outliers = iris.data[-5:]  # virginica will be considered as outliers
+    # lower the precision of the input data to check that we have an equivalence when
+    # making the computation in 32 and 64 bits.
+    X = np.concatenate([inliers, outliers], axis=0).astype(np.float32)
+
+    lof_32 = neighbors.LocalOutlierFactor(
+        algorithm=algorithm, novelty=novelty, contamination=contamination
+    )
+    X_32 = X.astype(np.float32, copy=True)
+    lof_32.fit(X_32)
+
+    lof_64 = neighbors.LocalOutlierFactor(
+        algorithm=algorithm, novelty=novelty, contamination=contamination
+    )
+    X_64 = X.astype(np.float64, copy=True)
+    lof_64.fit(X_64)
+
+    assert_allclose(lof_32.negative_outlier_factor_, lof_64.negative_outlier_factor_)
+
+    for method in ("score_samples", "decision_function", "predict", "fit_predict"):
+        if hasattr(lof_32, method):
+            y_pred_32 = getattr(lof_32, method)(X_32)
+            y_pred_64 = getattr(lof_64, method)(X_64)
+            assert_allclose(y_pred_32, y_pred_64)
+
```

jjerphan committed:
TST Add dedicated test for dtype equivalence

Authored-by: Guillaume Lemaitre <g.lemaitre58@gmail.com>

jjerphan committed:
TST Use proper atol

« _Les opticiens!_ »

jjerphan committed:
Merge branch 'main' into tst/test_lof-32bit

glemaitre reviewed (approved):
LGTM
